### PR TITLE
Layer system

### PR DESCRIPTION
**Pull request by serk12**
Signed-off-by: Marc Prat Masó <marc.prat.maso@estudiantat.upc.edu>

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Some examples:

### CODE REVIEW

File: avogadro/qtgui/layermodel.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        }
+        case Qt::ForegroundRole:
+          if (layer == static_cast<int>(getMoleculeLayer().activeLayer()))
+            return QVariant(QColor(Qt::red));
+          else
+            return QVariant(QColor(Qt::black));
+        default:
+          return QVariant();
+      }
+    } else if (idx.column() == (QTTY_COLUMNS - 5)) {
```
Comment (ghutchis) [inline avogadro/qtgui/layermodel.cpp:85]: I'd declare an enum  -- either in the header or here. That way you can have `(idx.column()  == LayerVisible)` and it makes it much easier to change / add  columns in the future.

e.g., `propertymodel.*`
https://github.com/OpenChemistry/avogadrolibs/pull/711/files

**Review by ghutchis — CHANGES_REQUESTED** [review]
Generally looks good - definitely needs some doc comments in headers. Also seems like a `.png` crept in.

Somewhere we need to document the source of the icons (e.g., README)

File: avogadro/core/layer.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  m_atomAndLayers.clear();
+  m_activeLayer = m_maxLayer = 0;
+}
+
+size_t Layer::activeLayer() const
+{
+  return m_activeLayer;
+}
+
+size_t Layer::maxLayer() const
```
[the remark this answers is not in the document]
Comment (serk12) [reply avogadro/core/layer.cpp:86]: `MaxLayer` is the known maximum, so it starts at 0. If you prefer I could rewrite it, but it's not a counting variable.

File: avogadro/core/layermanager.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * @class LayerData layermanager.h <avogadro/core/layermanager.h>
+ * @brief Interface to store layer data structure.
+ */
+struct LayerData
+{
+  LayerData(std::string save = "") { load(save); }
+
+  /** save custom data, base save should never be called */
+  virtual std::string save()
```
Comment (serk12) [inline avogadro/core/layermanager.h:33]: I'm not happy with this, is there anything in `Avogadro:IO` I could extend to? or any better idea?

File: avogadro/core/layer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghutchis) [reply avogadro/core/layer.cpp:86]: Okay. Still seems like there should be a `layerCount()` method to parallel atoms, bonds, residues, etc, but I guess that's `maxLayer() + 1`

File: avogadro/core/layermanager.h (hunk 3 above)
Comment (ghutchis) [reply avogadro/core/layermanager.h:33]: I mean, this is basically `serialize()` and `deserialize()` - you're not actually saving, you're just generating a string that represents the current layer configuration.

You can be sure the base method isn't called by making it a pure virtual method.

Comment (serk12) [reply avogadro/core/layermanager.h:33]: I can't use pure virtual method. I create `LayerData` objects in `cjsonformat.cpp` and then it's serialized in `pluginlayermanager.h` where I know the class type. But the renaming gives me some comfort .

File: avogadro/core/layermanager.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+ * @brief Interface to store layer data structure.
+ */
+struct LayerData
+{
+  LayerData(std::string save = "") { deserialize(save); }
+
+  /** save custom data, base save should never be called */
+  virtual std::string serialize()
+  {
+    assert(true);
```
Comment (ghutchis) [inline avogadro/core/layermanager.h:35]: Can we drop the `assert` if it's always true?

**Comment by ghutchis** [comment]
The Mac build failure needs `#include <string>` in `layermanager.h`

File: avogadro/core/layermanager.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+#include "avogadrocore.h"
+
+#include "array.h"
+#include "layer.h"
+
+#include <cassert>
+#include <map>
+#include <memory>
+#include <vector>
```
Comment (ghutchis) [inline avogadro/core/layermanager.h:17]: `#include <string>`

**Pull request merged**